_dependency_source, _cli, test: Do not override `pip.conf` unless explicitly specified via flags

philblckwd opened the pull request:
Refactor index-url and extra-index-url options to not override config defined in 'pip.conf' unless explicitly specified via flags.

Closes #563

woodruffw commented on pip_audit/_cli.py:
```diff
@@ -286,7 +285,6 @@ def _parser() -> argparse.ArgumentParser:  # pragma: no cover
         type=str,
         help="base URL of the Python Package Index; this should point to a repository compliant "
         "with PEP 503 (the simple repository API)",
-        default=PYPI_URL,
```
I think we want to keep this `default` in the CLI -- behaviorally it should remain the default, and having it appear as such in the `--help` render makes for a nicer UX 🙂

philblckwd replied:
So leaving this as the default means that it will always be explicitly passed to the `pip install` CLI, which results in the `index-url` specified in the `pip.conf` being ignored. I understand that it makes for a more friendly UX to specify as such, but would it be reasonable to convey that the url will be resolved by `pip` if not passed by the user explicitly?

If this is a deal breaker though, then I'll have to dig deeper into `pip` to determine how it is performing the `index-url` resolution.

woodruffw commented on pip_audit/_dependency_source/pyproject.py:
```diff
@@ -33,14 +28,20 @@ class PyProjectSource(DependencySource):
-        self, filename: Path, index_urls: list[str] = [PYPI_URL], state: AuditState = AuditState()
+        self,
+        filename: Path,
+        index_url: str = "",
```
Similar typestate nit here: maybe drop the default `""` here (or make it `PYPI_URL`), since we always expect a valid URL?

tetsuo-cpp replied:
Yep, I'd prefer this too.

woodruffw commented on pip_audit/_dependency_source/requirement.py:
```diff
@@ -47,7 +46,8 @@ def __init__(
         no_deps: bool = False,
         skip_editable: bool = False,
-        index_urls: list[str] = [PYPI_URL],
+        index_url: str = "",
```
Same as https://github.com/pypa/pip-audit/pull/565/files#r1145039056

woodruffw replied:
```suggestion
        index_url: str | None = None,
```

woodruffw commented:
Thanks a ton for opening this, @philblckwd! 

This looks great to me structurally; just a few small nitpicks 🙂

woodruffw commented:
(This also needs a `CHANGELOG.md entry. Feel free to do it yourself if you feel like it, otherwise I'll do it as a fixup after merge!)

tetsuo-cpp commented:
I'm also happy with where this is at. Once the help text is amended and those type state related comments are addressed, this should be good to go.

Thanks again for your work. We really appreciate it.

tetsuo-cpp commented:
@philblckwd One more thing. We have a check in CI to ensure that the help text that we display in the README is up to date. So once you've amended the help text for `--index-url`, you'll need to update the README.

philblckwd commented:
Per feedback:

- I updated arguments to use optional types
- I updated the help text, as well as the readme, to state the `index-url` will be resolved by pip if not specified
- I added an entry to the changelog
- I noticed unrelated lint and check-readme errors, so resolved those too

tetsuo-cpp reviewed (approved):
LGTM!

tetsuo-cpp commented:
I'll merge this shortly after I've done a bit of testing locally.

tetsuo-cpp commented:
Thanks @philblckwd!
